solver/types.go

```diff
@@ -56,6 +56,9 @@ type VertexOptions struct {
 	ExportCache  *bool
 	// WorkerConstraint
 	ProgressGroup *pb.ProgressGroup
+	// PinFunc takes a source.Identifier as the argument.
+	// Defined as func(interface{}) to avoid circular imports.
+	PinFunc func(context.Context, interface{}) error
```

Comment:
I don't think this should be saved in `VertexOptions` and then called during the resolve . Multiple builds with same LLB (but different policy) would share the `VertexOptions`.

Better would be to actually rewrite the LLB while it is getting loaded into the graph. So if `image(alpine)` is converted to `image(alpine@sha)` then it's vertex digest now also changes and matches another build that called `image(alpine@sha)` directly.

Reply:
Where would be the right place to rewrite the LLB?

Reply:
Somewhere in https://github.com/moby/buildkit/blob/68059406655653f5df4b35f5a4728d673920f166/solver/llbsolver/vertex.go#L223-L229 probably. You probably need to keep mapping of original digest to new digest after the `*pb.Op` has been rewritten so that the parent that has input with the old digest also gets modified to the new one.

Reply:
Updated like this

<details>

<p>

```diff
diff --git a/solver/llbsolver/vertex.go b/solver/llbsolver/vertex.go
index 4f36c2ed..b38edea1 100644
--- a/solver/llbsolver/vertex.go
+++ b/solver/llbsolver/vertex.go
@@ -1,6 +1,7 @@
 package llbsolver
 
 import (
+       "context"
        "fmt"
        "strings"
 
@@ -8,6 +9,7 @@ import (
        "github.com/moby/buildkit/solver"
        "github.com/moby/buildkit/solver/pb"
        "github.com/moby/buildkit/source"
+       srcpolapplier "github.com/moby/buildkit/sourcepolicy/applier"
        "github.com/moby/buildkit/util/entitlements"
        digest "github.com/opencontainers/go-digest"
        ocispecs "github.com/opencontainers/image-spec/specs-go/v1"
@@ -143,8 +145,8 @@ func (dpc *detectPrunedCacheID) Load(op *pb.Op, md *pb.OpMetadata, opt *solver.V
        return nil
 }
 
-func Load(def *pb.Definition, opts ...LoadOpt) (solver.Edge, error) {
-       return loadLLB(def, func(dgst digest.Digest, pbOp *pb.Op, load func(digest.Digest) (solver.Vertex, error)) (solver.Vertex, error) {
+func Load(ctx context.Context, def *pb.Definition, applier *srcpolapplier.Applier, opts ...LoadOpt) (solver.Edge, error) {
+       return loadLLB(ctx, def, applier, func(dgst digest.Digest, pbOp *pb.Op, load func(digest.Digest) (solver.Vertex, error)) (solver.Vertex, error) {
                opMetadata := def.Metadata[dgst]
                vtx, err := newVertex(dgst, pbOp, &opMetadata, load, opts...)
                if err != nil {
@@ -187,12 +189,13 @@ func newVertex(dgst digest.Digest, op *pb.Op, opMeta *pb.OpMetadata, load func(d
 
 // loadLLB loads LLB.
 // fn is executed sequentially.
-func loadLLB(def *pb.Definition, fn func(digest.Digest, *pb.Op, func(digest.Digest) (solver.Vertex, error)) (solver.Vertex, error)) (solver.Edge, error) {
+func loadLLB(ctx context.Context, def *pb.Definition, applier *srcpolapplier.Applier, fn func(digest.Digest, *pb.Op, func(digest.Digest) (solver.Vertex, error)) (solver.Vertex, error)) (solver.Edge, error) {
        if len(def.Def) == 0 {
                return solver.Edge{}, errors.New("invalid empty definition")
        }
 
        allOps := make(map[digest.Digest]*pb.Op)
+       mutatedDigests := make(map[digest.Digest]digest.Digest) // key: old, val: new
 
        var dgst digest.Digest
 
@@ -202,9 +205,31 @@ func loadLLB(def *pb.Definition, fn func(digest.Digest, *pb.Op, func(digest.Dige
                        return solver.Edge{}, errors.Wrap(err, "failed to parse llb proto op")
                }
                dgst = digest.FromBytes(dt)
+               mutated, err := applier.Apply(ctx, &op)
+               if err != nil {
+                       return solver.Edge{}, errors.Wrap(err, "failed to call source policy applier")
+               }
+               if mutated {
+                       dtMutated, err := op.Marshal()
+                       if err != nil {
+                               return solver.Edge{}, err
+                       }
+                       dgstMutated := digest.FromBytes(dtMutated)
+                       mutatedDigests[dgst] = dgstMutated
+                       dgst = dgstMutated
+               }
                allOps[dgst] = &op
        }
 
+       // Rewrite the input digests if some op was mutated by the source policy applier
+       for _, op := range allOps {
+               for _, inp := range op.Inputs {
+                       if mutatedDgst, ok := mutatedDigests[inp.Digest]; ok {
+                               inp.Digest = mutatedDgst
+                       }
+               }
+       }
+
        if len(allOps) < 2 {
                return solver.Edge{}, errors.Errorf("invalid LLB with %d vertexes", len(allOps))
        }
```

</p>

</details>